### PR TITLE
[Exporter.OTLP] Mark as nullable

### PR DESCRIPTION
**Pull request by Kielek**
Towards: #3958

## Changes

Mark whole project as a nullable.
I have a plan to include also test project to this PR, but it is to many changes.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* ~~[ ] Unit tests added/updated~~
* ~~[ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes~~
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/ExportClientResponse.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@ protected ExportClientResponse(bool success, DateTime deadlineUtc, Exception? ex
         this.DeadlineUtc = deadlineUtc;
     }
 
+    [MemberNotNullWhen(false, nameof(Exception))]
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/ExportClientResponse.cs:17]: This is interesting. A `bool` can never be `null`. Was there some warning that necessitated this?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OtlpExporterBuilder.cs [hunk 2]
```diff
@@ -121,10 +116,6 @@ internal sealed class OtlpExporterBuilder
 
     private static void BindConfigurationToOptions(IServiceCollection services, string name, IConfiguration configuration)
     {
-        Debug.Assert(services != null, "services was null");
-        Debug.Assert(!string.IsNullOrEmpty(name), "name was null or empty");
-        Debug.Assert(configuration != null, "configuration was null");
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OtlpExporterBuilder.cs:126]: I feel like we should keep all of the `Debug.Assert`s being removed on this PR. Those are basically sanity checking that what we declare should not be `null` does in fact have a value. Really nullable annotations are just hints. They don't enforce anything and devs can still make mistakes/decide to override things with `!`. IMO, it is best practice to always `Debug.Assert` those assumptions.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/ExportClientResponse.cs (hunk 1 above)
Comment (Kielek) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/ExportClientResponse.cs:17]: This mean that Exception property will be not null when this property is false.
Yes, there were warning in couple of places.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OtlpExporterBuilder.cs (hunk 2 above)
Comment (Kielek) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OtlpExporterBuilder.cs:126]: I can revert this. At least for.NET Fx I will need to put `!` for usages. Compiler does not undertand it correctly.
Alternative option is to use Guard, but it can be more expensive in execution.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/ExportClientResponse.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/ExportClientResponse.cs:17]: Ah got it this is about `Exception` property not `Success`. Makes sense.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OtlpExporterBuilder.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OtlpExporterBuilder.cs:126]: Ya I know it's annoying putting all the `!`s just for .NET Fx \ .NET Standard. Definitely prefer adding the `!`s over calling into `Guard` though.

We could drop something like this into `Common.props`:

```xml
	<PropertyGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
                <!-- Disable nullable warnings on older TFMs -->
		<NoWarn>$(NoWarn);8600;8601;8602;8603;8604;8618</NoWarn>
	</PropertyGroup>
```

The idea is to turn off nullable analysis warnings for older TFMs where the IDE/compiler are half-baked. That would eliminate the need for `!`s just for the older things. I didn't do this before because we had some projects which only targeted say `netstandard2.0` but AFAIK everything now has at least one modern `netX.0` target to cover the analysis.

Comment (Kielek) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Builder/OtlpExporterBuilder.cs:126]: In this PR I have reverted Debug assetions without disabling analysis for .NET Fx/ .NET Standard. I think it will be better to do in separate requests. There is a good tooling to automatically remoing unneeded null forgiving operator `!` globally in project/solution. 

See 76a1c9d4b898e894c38439b5429a54aaff3266a3

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ResourceExtensions.cs [hunk 3]
```diff
@@ -17,7 +17,7 @@ public static OtlpResource.Resource ToOtlpResource(this Resource resource)
 
         foreach (KeyValuePair<string, object> attribute in resource.Attributes)
         {
-            OtlpTagWriter.Instance.TryWriteTag(ref processResourceAttributes, attribute);
+            OtlpTagWriter.Instance.TryWriteTag(ref processResourceAttributes, attribute.Key, attribute.Value);
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ResourceExtensions.cs:20]: Is this related to nullable analysis? I don't have an issue with the change, just don't understand why it would be needed 😄

**Review by CodeBlanch — APPROVED** [review]
LGTM

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ResourceExtensions.cs (hunk 3 above)
Comment (Kielek) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ResourceExtensions.cs:20]: TryWriteTag is accepting `KeyValuePair<string, object?>`. Rosources returing `KeyValuePair<string, object>` (objest is non-nullable). The cheapest way I have found, without repacking object was to create overload with two separate parameters - key and value. Happy to change if you see better option for fixes.
